README
DirectFB2-gles2
===============

DirectFB2-gles2 contains the OpenGL ES 2.0 GFX driver for DirectFB2.


ChangeLog
commit 72d55e0155abd60d78767bf10fb01132d961736a
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Sun Mar 26 08:42:55 2023 +0200

    Check GL_RENDERER string in driver_probe()

commit 64e273aef5a45023cb56cd41080c09d242dd67eb
Merge: 209b43e 0033ae3
Author: DirectFB2 <80045286+directfb2@users.noreply.github.com>
Date:   Thu Mar 23 17:20:30 2023 +0100

    Merge pull request #5 from caramelli/master
    
    Add pkg-config file (for static build)

commit 0033ae3f5531ed1f0d9bf6803f36c93607d1fd43
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Thu Mar 23 17:17:35 2023 +0100

    Update ChangeLog

commit be3c653d6e566310a4169101bc10f571003327bc
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Thu Mar 23 17:17:27 2023 +0100

    Add pkg-config file (for static build)

commit 209b43eaa71db59e9d0a671ec133b34d435e3487
Merge: 0f8eb34 bfe5599
Author: DirectFB2 <80045286+directfb2@users.noreply.github.com>
Date:   Wed Jan 11 01:01:56 2023 +0100

    Merge pull request #4 from caramelli/master
    
    Add support for layer rotation

commit bfe55996872e782a099087af55fb13dcfcb35258
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Jan 11 00:59:08 2023 +0100

    Update ChangeLog

commit 3ecb795f89178611fdb45fb1691ae6e0e54551f2
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Jan 11 00:58:53 2023 +0100

    Add support for layer rotation

commit 0f8eb34ba297e9cb6e4aecc1dbdab13f5ee6d400
Merge: fe03aa2 ec50cf0
Author: DirectFB2 <80045286+directfb2@users.noreply.github.com>
Date:   Fri Jan 6 22:12:36 2023 +0100

    Merge pull request #3 from caramelli/master
    
    Add hardware acceleration support for 90 and 270 degree rotations

commit ec50cf0196fa463acecb247228dc6fd21393b3ff
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Fri Jan 6 22:11:25 2023 +0100

    Update ChangeLog

commit 6ac7bf903d9620f766b6af30892ec83974e23c5c
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Fri Jan 6 22:11:20 2023 +0100

    Add hardware acceleration support for 90 and 270 degree rotations

commit fe03aa2d387ef62fc91197a2cdb4aeff57f1fc4f
Merge: db73f1c 3ca2634
Author: DirectFB2 <80045286+directfb2@users.noreply.github.com>
Date:   Fri Nov 11 14:54:44 2022 +0100

    Merge pull request #2 from caramelli/master
    
    fragment shaders: Add precision mediump float

commit 3ca2634c19a5f08a614547576ce3f9d5299ef542
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Fri Nov 11 14:54:07 2022 +0100

    Update ChangeLog

commit eb7049dba355c9a4abd46158da6f5d1b1bbbc0d0
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Fri Nov 11 14:53:56 2022 +0100

    fragment shaders: Add precision mediump float

commit db73f1ccb7b8355285b70805e7915a7f9288445b
Merge: 6b6cd66 87d85d3
Author: DirectFB2 <80045286+directfb2@users.noreply.github.com>
Date:   Wed Nov 9 17:24:56 2022 +0100

    Merge pull request #1 from caramelli/master
    
    OpenGL ES 2.0 GFX driver

commit 87d85d3d8e161968babbdffa8234b0b7eb364242
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Nov 9 17:23:05 2022 +0100

    Add ChangeLog

commit e0146eedae1e027624d3b256d7e93e5179ab80ae
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Nov 9 17:22:48 2022 +0100

    OpenGL ES 2.0 GFX driver

commit 6b6cd665a4ce106da0631fd798cdd64b60f5a467
Author: Nicolas Caramelli <caramelli.devel@gmail.com>
Date:   Wed Nov 9 17:20:30 2022 +0100

    Initial revision


gles2_shaders.h
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

/* Transform input 2D positions "dfbPos" by scale and offset to get GLES clip coordinates. */
static const char *draw_vert_src = "                                     \
attribute vec2 dfbPos;                                                   \
uniform   vec3 dfbScale;                                                 \
uniform   mat3 dfbRotMatrix;                                             \
                                                                         \
void main(void)                                                          \
{                                                                        \
     vec3 pos;                                                           \
     pos.x = dfbScale.x * dfbPos.x - 1.0;                                \
     pos.y = dfbScale.y * dfbPos.y + dfbScale.z;                         \
     pos.z = 0.0;                                                        \
     gl_Position = vec4(dfbRotMatrix * pos, 1.0);                        \
}";

/* Transform input 2D positions "dfbPos" by the render options matrix before transforming to GLES clip coordinates. */
static const char *draw_mat_vert_src = "                                 \
attribute vec2 dfbPos;                                                   \
uniform   mat3 dfbMVPMatrix;                                             \
uniform   mat3 dfbROMatrix;                                              \
                                                                         \
void main(void)                                                          \
{                                                                        \
     vec3 pos = dfbMVPMatrix * dfbROMatrix * vec3(dfbPos, 0.0);          \
     gl_Position = vec4(pos.x, pos.y, 0.0, 1.0);                         \
}";

/* Draw fragment in a constant color. */
static const char *draw_frag_src = "                                     \
precision mediump float;                                                 \
                                                                         \
uniform vec4 dfbColor;                                                   \
                                                                         \
void main(void)                                                          \
{                                                                        \
     gl_FragColor = dfbColor;                                            \
}";

/* This is the same as draw_vert_src with the addition of texture coords "dfbUV". */
static const char *blit_vert_src = "                                     \
attribute vec2 dfbPos;                                                   \
attribute vec2 dfbUV;                                                    \
uniform   vec3 dfbScale;                                                 \
uniform   mat3 dfbRotMatrix;                                             \
uniform   vec2 dfbTexScale;                                              \
varying   vec2 varTexCoord;                                              \
                                                                         \
void main(void)                                                          \
{                                                                        \
     vec3 pos;                                                           \
     pos.x = dfbScale.x * dfbPos.x - 1.0;                                \
     pos.y = dfbScale.y * dfbPos.y + dfbScale.z;                         \
     pos.z = 0.0;                                                        \
     gl_Position = vec4(dfbRotMatrix * pos, 1.0);                        \
     varTexCoord.s = dfbTexScale.x * dfbUV.x;                            \
     varTexCoord.t = dfbTexScale.y * dfbUV.y;                            \
}";

/* This is the same as draw_mat_vert_src with the addition of texture coords "dfbUV". */
static const char *blit_mat_vert_src = "                                 \
attribute vec2 dfbPos;                                                   \
attribute vec2 dfbUV;                                                    \
uniform   mat3 dfbMVPMatrix;                                             \
uniform   mat3 dfbROMatrix;                                              \
uniform   vec2 dfbTexScale;                                              \
varying   vec2 varTexCoord;                                              \
                                                                         \
void main(void)                                                          \
{                                                                        \
     vec3 pos = dfbMVPMatrix * dfbROMatrix * vec3(dfbPos, 0.0);          \
     gl_Position = vec4(pos.x, pos.y, 0.0, 1.0);                         \
     varTexCoord.s = dfbTexScale.x * dfbUV.x;                            \
     varTexCoord.t = dfbTexScale.y * dfbUV.y;                            \
}";

/* Sample texture. */
static const char *blit_frag_src = "                                     \
precision mediump float;                                                 \
                                                                         \
uniform sampler2D dfbSampler;                                            \
varying vec2      varTexCoord;                                           \
                                                                         \
void main(void)                                                          \
{                                                                        \
     gl_FragColor = texture2D(dfbSampler, varTexCoord);                  \
}";

/* Sample texture and modulate by static color. */
static const char *blit_color_frag_src = "                               \
precision mediump float;                                                 \
                                                                         \
uniform sampler2D dfbSampler;                                            \
uniform vec4      dfbColor;                                              \
varying vec2      varTexCoord;                                           \
                                                                         \
void main(void)                                                          \
{                                                                        \
     gl_FragColor = texture2D(dfbSampler, varTexCoord) * dfbColor;       \
}";

/* Apply source color keying. */
static const char *blit_colorkey_frag_src = "                            \
precision mediump float;                                                 \
                                                                         \
uniform sampler2D dfbSampler;                                            \
uniform vec4      dfbColor;                                              \
uniform ivec3     dfbColorkey;                                           \
varying vec2      varTexCoord;                                           \
                                                                         \
void main(void)                                                          \
{                                                                        \
     vec4 c = texture2D(dfbSampler, varTexCoord);                        \
     int  r = int(c.r * 255.0 + 0.5);                                    \
     int  g = int(c.g * 255.0 + 0.5);                                    \
     int  b = int(c.b * 255.0 + 0.5);                                    \
     if (r == dfbColorkey.x && g == dfbColorkey.y && b == dfbColorkey.z) \
          discard;                                                       \
     gl_FragColor = c * dfbColor;                                        \
}";

/* Perform an alpha pre-multiply of source frag color with source frag alpha after sampling and modulation. */
static const char *blit_premultiply_frag_src = "                         \
precision mediump float;                                                 \
                                                                         \
uniform sampler2D dfbSampler;                                            \
uniform vec4      dfbColor;                                              \
varying vec2      varTexCoord;                                           \
                                                                         \
void main(void)                                                          \
{                                                                        \
     gl_FragColor  = texture2D(dfbSampler, varTexCoord);                 \
     gl_FragColor *= dfbColor;                                           \
     gl_FragColor.rgb *= gl_FragColor.a;                                 \
}";


gles2_gfxdriver.h
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#ifndef __GLES2_GFXDRIVER_H__
#define __GLES2_GFXDRIVER_H__

#include <GLES2/gl2.h>

/**********************************************************************************************************************/

typedef enum {
     GLES2VA_POSITIONS = 0,
     GLES2VA_TEXCOORDS = 1
} GLES2VertexAttribs;

typedef enum {
     NONE        = 0x00000000,

     DESTINATION = 0x00000001,
     CLIP        = 0x00000002,
     MATRIX      = 0x00000004,

     COLOR_DRAW  = 0x00000010,
     COLORKEY    = 0x00000020,

     SOURCE      = 0x00000100,
     COLOR_BLIT  = 0x00000200,

     BLENDING    = 0x00010000,

     ALL         = 0x00010337
} GLES2ValidationFlags;

typedef struct {
     GLuint                obj;          /* the program object */
     GLint                 dfbScale;     /* location of scale factors for clipped coordinates */
     GLint                 dfbRotMatrix; /* location of layer rotation matrix */
     GLint                 dfbROMatrix;  /* location of render options matrix */
     GLint                 dfbMVPMatrix; /* location of model-view-projection matrix */
     GLint                 dfbColor;     /* location of global RGBA color */
     GLint                 dfbColorkey;  /* location of colorkey RGB color */
     GLint                 dfbTexScale;  /* location of scale factors for normalized tex coordinates */
     char                 *name;         /* program object name for debugging */
     GLES2ValidationFlags  flags;        /* validation flags */
} GLES2ProgramInfo;

typedef enum {
     DRAW,
     DRAW_MAT,
     BLIT,
     BLIT_MAT,
     BLIT_COLOR,
     BLIT_COLOR_MAT,
     BLIT_COLORKEY,
     BLIT_COLORKEY_MAT,
     BLIT_PREMULTIPLY,
     BLIT_PREMULTIPLY_MAT,
     NUM_PROGRAMS,
     INVALID_PROGRAM
} GLES2ProgramIndex;

typedef struct {
     DFBSurfaceBlittingFlags blittingflags; /* blitting flags */
     float                   aspect;        /* layer aspect scaling */
     int                     rotation;      /* layer rotation */
} GLES2DriverData;

typedef struct {
     GLES2ProgramInfo  progs[NUM_PROGRAMS]; /* program info */
     GLES2ProgramIndex prog_index;          /* current program in use */
} GLES2DeviceData;

#endif


gles2_gfxdriver.c
/*
   This file is part of DirectFB.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include <core/graphics_driver.h>
#include <misc/conf.h>

#include "gles2_gfxdriver.h"
#include "gles2_shaders.h"

D_DEBUG_DOMAIN( GLES2_Driver, "GLES2/Driver", "OpenGL ES 2.0 Driver" );

DFB_GRAPHICS_DRIVER( gles2 )

/**********************************************************************************************************************/

extern const GraphicsDeviceFuncs gles2GraphicsDeviceFuncs;

static DFBResult
init_shader( GLuint      prog_obj,
             const char *prog_src,
             GLenum      type )
{
     GLuint  shader;
     GLint   status;
     GLint   log_length;
     GLchar *log;

     /* Create the shader. */
     shader = glCreateShader( type );
     if (!shader)
          return DFB_FAILURE;

     glShaderSource( shader, 1, &prog_src, NULL );

     /* Compile the shader. */
     glCompileShader( shader );

     glGetShaderiv( shader, GL_COMPILE_STATUS, &status );
     if (!status) {
          glGetShaderiv( shader, GL_INFO_LOG_LENGTH, &log_length );
          log = D_MALLOC( log_length );
          if (!log)
               return D_OOM();

          glGetShaderInfoLog( shader, log_length, NULL, log );
          D_ERROR( "GLES2/Driver: Failed to compile shader:\n%s\n", log );

          D_FREE( log );
          glDeleteShader( shader );

          return DFB_FAILURE;
     }

     /* Attach shader to program object. */
     glAttachShader( prog_obj, shader );

     /* Mark shader for deletion (deletion will not occur until shader is detached). */
     glDeleteShader( shader );

     return DFB_OK;
}

static GLuint
init_program( const char *vert_prog_src,
              const char *frag_prog_src,
              DFBBoolean  texcoords )
{
     GLuint  prog_obj;
     GLuint  shaders[2];
     GLint   status;
     GLint   log_length;
     char   *log;

     /* Create the program object. */
     prog_obj = glCreateProgram();
     if (!prog_obj)
          return 0;

     /* Create the vertex shader. */
     if (init_shader( prog_obj, vert_prog_src, GL_VERTEX_SHADER )) {
          D_ERROR( "GLES2/Driver: Failed to create vertex shader!\n" );
          return 0;
     }

     /* Create the fragment shader. */
     if (init_shader( prog_obj, frag_prog_src, GL_FRAGMENT_SHADER )) {
          D_ERROR( "GLES2/Driver: Failed to create fragment shader!\n" );
          return 0;
     }

     /* Bind vertex positions to "dfbPos". */
     glBindAttribLocation( prog_obj, GLES2VA_POSITIONS, "dfbPos" );

     /* Bind vertex texture coords to "dfbUV". */
     if (texcoords)
          glBindAttribLocation( prog_obj, GLES2VA_TEXCOORDS, "dfbUV" );

     /* Link the program object. */
     glLinkProgram( prog_obj );

     glGetProgramiv( prog_obj, GL_LINK_STATUS, &status );
     if (!status) {
          glGetProgramiv( prog_obj, GL_INFO_LOG_LENGTH, &log_length );
          log = D_MALLOC( log_length );
          if (!log)
               return 0;

          glGetProgramInfoLog( prog_obj, log_length, NULL, log );
          D_ERROR( "GLES2/Driver: Failed to link program:\n%s\n", log );

          D_FREE( log );
          return 0;
     }

     /* No need for shaders anymore. */
     glGetAttachedShaders( prog_obj, 2, NULL, shaders );
     glDetachShader( prog_obj, shaders[0] );
     glDetachShader( prog_obj, shaders[1] );

     return prog_obj;
}

/**********************************************************************************************************************/

static int
driver_probe()
{
     return glGetString( GL_RENDERER ) ? 1 : 0;
}

static void
driver_get_info( GraphicsDriverInfo *driver_info )
{
     driver_info->version.major = 0;
     driver_info->version.minor = 1;

     snprintf( driver_info->name,   DFB_GRAPHICS_DRIVER_INFO_NAME_LENGTH,   "GLES2" );
     snprintf( driver_info->vendor, DFB_GRAPHICS_DRIVER_INFO_VENDOR_LENGTH, "DirectFB" );

     driver_info->driver_data_size = sizeof(GLES2DriverData);
     driver_info->device_data_size = sizeof(GLES2DeviceData);
}

static DFBResult
driver_init_driver( GraphicsDeviceFuncs *funcs,
                    void                *driver_data,
                    void                *device_data,
                    CoreDFB             *core )
{
     D_DEBUG_AT( GLES2_Driver, "%s()\n", __FUNCTION__ );

     dfb_config->font_format = DSPF_ARGB;

     *funcs = gles2GraphicsDeviceFuncs;

     return DFB_OK;
}

static DFBResult
driver_init_device( GraphicsDeviceInfo *device_info,
                    void               *driver_data,
                    void               *device_data )
{
     GLES2DeviceData *dev = device_data;
     GLuint           prog_obj;
     int              i;

     D_DEBUG_AT( GLES2_Driver, "%s()\n", __FUNCTION__ );

     /* Fill device information. */
     snprintf( device_info->name,   DFB_GRAPHICS_DEVICE_INFO_NAME_LENGTH,   "%s", glGetString( GL_RENDERER ) );
     snprintf( device_info->vendor, DFB_GRAPHICS_DEVICE_INFO_VENDOR_LENGTH, "OpenGL ES" );
     device_info->caps.flags    = CCF_CLIPPING | CCF_RENDEROPTS;
     device_info->caps.accel    = DFXL_FILLRECTANGLE | DFXL_DRAWRECTANGLE | DFXL_DRAWLINE | DFXL_FILLTRIANGLE |
                                  DFXL_BLIT          | DFXL_STRETCHBLIT;
     device_info->caps.blitting = DSBLIT_BLEND_ALPHACHANNEL | DSBLIT_BLEND_COLORALPHA | DSBLIT_COLORIZE         |
                                  DSBLIT_SRC_COLORKEY       | DSBLIT_SRC_PREMULTIPLY  | DSBLIT_SRC_PREMULTCOLOR |
                                  DSBLIT_ROTATE180          | DSBLIT_ROTATE90         | DSBLIT_ROTATE270;
     device_info->caps.drawing  = DSDRAW_BLEND | DSDRAW_SRC_PREMULTIPLY;

     /* Initialize program information. */
     for (i = 0; i < NUM_PROGRAMS; i++) {
          dev->progs[i].obj          =  0;
          dev->progs[i].dfbScale     = -1;
          dev->progs[i].dfbRotMatrix = -1;
          dev->progs[i].dfbROMatrix  = -1;
          dev->progs[i].dfbMVPMatrix = -1;
          dev->progs[i].dfbColor     = -1;
          dev->progs[i].dfbColorkey  = -1;
          dev->progs[i].dfbTexScale  = -1;
          dev->progs[i].flags        =  NONE;
          dev->progs[i].name         = "invalid";
     }

     /*
      * The draw program transforms a vertex by the current model-view-projection matrix,
      * applies a constant color to all fragments.
      */

     prog_obj = init_program( draw_vert_src, draw_frag_src, DFB_FALSE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create draw program!\n" );
          goto fail;
     }

     dev->progs[DRAW].obj          = prog_obj;
     dev->progs[DRAW].name         = "draw";
     dev->progs[DRAW].dfbColor     = glGetUniformLocation( dev->progs[DRAW].obj, "dfbColor" );
     dev->progs[DRAW].dfbScale     = glGetUniformLocation( dev->progs[DRAW].obj, "dfbScale" );
     dev->progs[DRAW].dfbRotMatrix = glGetUniformLocation( dev->progs[DRAW].obj, "dfbRotMatrix" );

     prog_obj = init_program( draw_mat_vert_src, draw_frag_src, DFB_FALSE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create draw_mat program!\n" );
          goto fail;
     }

     dev->progs[DRAW_MAT].obj          = prog_obj;
     dev->progs[DRAW_MAT].name         = "draw_mat";
     dev->progs[DRAW_MAT].dfbColor     = glGetUniformLocation( dev->progs[DRAW_MAT].obj, "dfbColor" );
     dev->progs[DRAW_MAT].dfbROMatrix  = glGetUniformLocation( dev->progs[DRAW_MAT].obj, "dfbROMatrix" );
     dev->progs[DRAW_MAT].dfbMVPMatrix = glGetUniformLocation( dev->progs[DRAW_MAT].obj, "dfbMVPMatrix" );

     /*
      * The blit program transforms a vertex by the current model-view-projection matrix,
      * applies texture sample colors to fragments.
      */

     prog_obj = init_program( blit_vert_src, blit_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit program!\n" );
          goto fail;
     }

     dev->progs[BLIT].obj          = prog_obj;
     dev->progs[BLIT].name         = "blit";
     dev->progs[BLIT].dfbScale     = glGetUniformLocation( dev->progs[BLIT].obj, "dfbScale" );
     dev->progs[BLIT].dfbRotMatrix = glGetUniformLocation( dev->progs[BLIT].obj, "dfbRotMatrix" );
     dev->progs[BLIT].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT].obj, "dfbTexScale" );

     prog_obj = init_program( blit_mat_vert_src, blit_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_mat program!\n" );
          goto fail;
     }

     dev->progs[BLIT_MAT].obj          = prog_obj;
     dev->progs[BLIT_MAT].name         = "blit_mat";
     dev->progs[BLIT_MAT].dfbROMatrix  = glGetUniformLocation( dev->progs[BLIT_MAT].obj, "dfbROMatrix" );
     dev->progs[BLIT_MAT].dfbMVPMatrix = glGetUniformLocation( dev->progs[BLIT_MAT].obj, "dfbMVPMatrix" );
     dev->progs[BLIT_MAT].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_MAT].obj, "dfbTexScale" );

     /*
      * The blit_color program transforms a vertex by the current model-view-projection matrix,
      * applies texture sample colors to fragments, and modulates the colors with a static color.
      * Modulation is effectively disabled by setting the static color components to 1.0.
      */

     prog_obj = init_program( blit_vert_src, blit_color_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_color program!\n" );
          goto fail;
     }

     dev->progs[BLIT_COLOR].obj          = prog_obj;
     dev->progs[BLIT_COLOR].name         = "blit_color";
     dev->progs[BLIT_COLOR].dfbColor     = glGetUniformLocation( dev->progs[BLIT_COLOR].obj, "dfbColor" );
     dev->progs[BLIT_COLOR].dfbScale     = glGetUniformLocation( dev->progs[BLIT_COLOR].obj, "dfbScale" );
     dev->progs[BLIT_COLOR].dfbRotMatrix = glGetUniformLocation( dev->progs[BLIT_COLOR].obj, "dfbRotMatrix" );
     dev->progs[BLIT_COLOR].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_COLOR].obj, "dfbTexScale" );

     prog_obj = init_program( blit_mat_vert_src, blit_color_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_color_mat program!\n" );
          goto fail;
     }

     dev->progs[BLIT_COLOR_MAT].obj          = prog_obj;
     dev->progs[BLIT_COLOR_MAT].name         = "blit_color_mat";
     dev->progs[BLIT_COLOR_MAT].dfbColor     = glGetUniformLocation( dev->progs[BLIT_COLOR_MAT].obj, "dfbColor" );
     dev->progs[BLIT_COLOR_MAT].dfbROMatrix  = glGetUniformLocation( dev->progs[BLIT_COLOR_MAT].obj, "dfbROMatrix" );
     dev->progs[BLIT_COLOR_MAT].dfbMVPMatrix = glGetUniformLocation( dev->progs[BLIT_COLOR_MAT].obj, "dfbMVPMatrix" );
     dev->progs[BLIT_COLOR_MAT].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_COLOR_MAT].obj, "dfbTexScale" );

     /*
      * The blit_colorkey program does the same as the blit program with the addition of source color keying.
      * Shaders don't have access to destination pixels, so color keying can only apply to the source.
      */

     prog_obj = init_program( blit_vert_src, blit_colorkey_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_colorkey program!\n" );
          goto fail;
     }

     dev->progs[BLIT_COLORKEY].obj          = prog_obj;
     dev->progs[BLIT_COLORKEY].name         = "blit_colorkey";
     dev->progs[BLIT_COLORKEY].dfbColor     = glGetUniformLocation( dev->progs[BLIT_COLORKEY].obj, "dfbColor" );
     dev->progs[BLIT_COLORKEY].dfbScale     = glGetUniformLocation( dev->progs[BLIT_COLORKEY].obj, "dfbScale" );
     dev->progs[BLIT_COLORKEY].dfbRotMatrix = glGetUniformLocation( dev->progs[BLIT_COLORKEY].obj, "dfbRotMatrix" );
     dev->progs[BLIT_COLORKEY].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_COLORKEY].obj, "dfbTexScale" );
     dev->progs[BLIT_COLORKEY].dfbColorkey  = glGetUniformLocation( dev->progs[BLIT_COLORKEY].obj, "dfbColorkey" );

     prog_obj = init_program( blit_mat_vert_src, blit_colorkey_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_colorkey_mat program!\n" );
          goto fail;
     }

     dev->progs[BLIT_COLORKEY_MAT].obj          = prog_obj;
     dev->progs[BLIT_COLORKEY_MAT].name         = "blit_colorkey_mat";
     dev->progs[BLIT_COLORKEY_MAT].dfbColor     = glGetUniformLocation( dev->progs[BLIT_COLORKEY_MAT].obj, "dfbColor" );
     dev->progs[BLIT_COLORKEY_MAT].dfbROMatrix  = glGetUniformLocation( dev->progs[BLIT_COLORKEY_MAT].obj, "dfbROMatrix" );
     dev->progs[BLIT_COLORKEY_MAT].dfbMVPMatrix = glGetUniformLocation( dev->progs[BLIT_COLORKEY_MAT].obj, "dfbMVPMatrix" );
     dev->progs[BLIT_COLORKEY_MAT].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_COLORKEY_MAT].obj, "dfbTexScale" );
     dev->progs[BLIT_COLORKEY_MAT].dfbColorkey  = glGetUniformLocation( dev->progs[BLIT_COLORKEY_MAT].obj, "dfbColorkey" );

     /*
      * The blit_premultiply program does the same as the blit program with the addition of pre-multiplication
      * of the source frag color by the source frag alpha.
      * Shaders don't have access to destination pixels, so pre-multiplication can only apply to the source.
      */

     prog_obj = init_program( blit_vert_src, blit_premultiply_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_premultiply program!\n" );
          goto fail;
     }

     dev->progs[BLIT_PREMULTIPLY].obj          = prog_obj;
     dev->progs[BLIT_PREMULTIPLY].name         = "blit_premultiply";
     dev->progs[BLIT_PREMULTIPLY].dfbColor     = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY].obj, "dfbColor" );
     dev->progs[BLIT_PREMULTIPLY].dfbScale     = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY].obj, "dfbScale" );
     dev->progs[BLIT_PREMULTIPLY].dfbRotMatrix = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY].obj, "dfbRotMatrix" );
     dev->progs[BLIT_PREMULTIPLY].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY].obj, "dfbTexScale" );

     prog_obj = init_program( blit_mat_vert_src, blit_premultiply_frag_src, DFB_TRUE );
     if (!prog_obj) {
          D_ERROR( "GLES2/Driver: Failed to create blit_premultiply_mat program!\n" );
          goto fail;
     }

     dev->progs[BLIT_PREMULTIPLY_MAT].obj          = prog_obj;
     dev->progs[BLIT_PREMULTIPLY_MAT].name         = "blit_premultiply_mat";
     dev->progs[BLIT_PREMULTIPLY_MAT].dfbColor     = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY_MAT].obj, "dfbColor" );
     dev->progs[BLIT_PREMULTIPLY_MAT].dfbROMatrix  = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY_MAT].obj, "dfbROMatrix" );
     dev->progs[BLIT_PREMULTIPLY_MAT].dfbMVPMatrix = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY_MAT].obj, "dfbMVPMatrix" );
     dev->progs[BLIT_PREMULTIPLY_MAT].dfbTexScale  = glGetUniformLocation( dev->progs[BLIT_PREMULTIPLY_MAT].obj, "dfbTexScale" );

     /* No program is used yet. */
     dev->prog_index = INVALID_PROGRAM;

     return DFB_OK;

fail:
     /* Delete program object, shaders are automatically detached, then deleted (they have been marked for deletion),
        program object with a value of 0 is automatically ignored. */
     for (i = 0; i < NUM_PROGRAMS; i++)
          glDeleteProgram(dev->progs[i].obj);

     return DFB_INIT;
}

static void
driver_close_device( void *driver_data,
                     void *device_data )
{
     D_DEBUG_AT( GLES2_Driver, "%s()\n", __FUNCTION__ );
}

static void
driver_close_driver( void *driver_data )
{
     D_DEBUG_AT( GLES2_Driver, "%s()\n", __FUNCTION__ );
}
